### PR TITLE
algebra: Refactor implementation to be generic over curve choice

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
This PR refactors the `algebra` module to be generic over curve choice; I'll come back through and fix comments, tests, etc after refactoring the rest of the crate

### Testing
- Tests will fail while refactor is incomplete

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -96,6 +96,7 @@ tracing = { version = "0.1", features = ["log"] }
 zeroize = "1.3"
 
 [dev-dependencies]
+ark-curve25519 = "0.4"
```
Comment (akirillo) [inline Cargo.toml:99]: note to self: why do we introduce a curve-specific dependency?

File: src/algebra/curve.rs [hunk 2]
```diff
@@ -158,44 +128,55 @@ impl StarkPoint {
...
-    pub fn from_bytes(bytes: &[u8]) -> Result<StarkPoint, SerializationError> {
-        let point = StarkPointInner::deserialize_compressed(bytes)?;
-        Ok(StarkPoint(point))
+    pub fn from_bytes(bytes: &[u8]) -> Result<CurvePoint<C>, SerializationError> {
+        let point = C::deserialize_compressed(bytes)?;
+        Ok(CurvePoint(point))
+    }
+
+    /// Get the number of bytes needed to represent a point, this is exactly the number of bytes
+    /// for one base field element, as we can simply use the x-coordinate and set a high bit for the `y`
```
Comment (akirillo) [inline src/algebra/curve.rs:137]: Is this true for all different types of curves?

File: src/algebra/curve.rs [hunk 3]
```diff
@@ -873,102 +867,85 @@ impl StarkPointResult {
...
-    };
+    use crate::algebra::test_helper::{random_point, TestCurve};
 
     use super::*;
     /// Test that the generators are the same between the two curve representations
     #[test]
     fn test_generators() {
-        let generator_1 = StarkPoint::generator();
-        let generator_2 = ProjectivePoint::from_affine_point(&GENERATOR);
+        // let generator_1 = CurvePoint::generator();
```
Comment (akirillo) [inline src/algebra/curve.rs:877]: are these commented-out tests dead code to be removed? or will they be implemented later?

Comment (akirillo) [reply src/algebra/curve.rs:877]: nvm, saw pr description, github buggy and won't let me delete this

File: src/algebra/scalar.rs [hunk 4]
```diff
@@ -155,60 +146,60 @@ impl<'de> Deserialize<'de> for Scalar {
 
 // === Addition === //
 
-/// A type alias for a result that resolves to a `Scalar`
-pub type ScalarResult = ResultHandle<Scalar>;
-/// A type alias for a result that resolves to a batch of `Scalar`s
-pub type BatchScalarResult = ResultHandle<Vec<Scalar>>;
-impl ScalarResult {
+/// A type alias for a result that resolves to a `Scalar<C>`
+pub type ScalarResult<C> = ResultHandle<C, Scalar<C>>;
```
Comment (akirillo) [inline src/algebra/scalar.rs:150]: getting build errors here about `ResultHandle` only accepting 1 generic param, I assume that's a change that's just not in this branch?

**Review by akirillo — APPROVED** [review]
lgtm

File: Cargo.toml (hunk 1 above)
Comment (akirillo) [reply Cargo.toml:99]: not letting me delete smh

File: src/algebra/curve.rs (hunk 2 above)
Comment (joeykraut) [reply src/algebra/curve.rs:137]: Not universally, it is true for basically any curve that could be used here.

File: src/algebra/scalar.rs (hunk 4 above)
Comment (joeykraut) [reply src/algebra/scalar.rs:150]: yeah

**Pull request merged**